doc/guides/V8-api-for-node.md

```diff
@@ -0,0 +1,210 @@
+  - can get Environment: Environment* env = Environment::GetCurrent(args);
+  - can get args using 0-based index
+  - returns Local<Value>
+  - has a .Length(), access past length returns a Local<Value> where value is
+    Undefined
+  - has a number of Is*() predicates which check exact type of Value, and
+    (mostly) do NOT consider possible conversions:
+    - {then: ()=>{}} is not considered a Promise,
+    - `1` is not considered `true`,
+    - `null` is not an `Object` (!),
+    - new String() is a StringObject (not a String),
+    - 3 is a Int32 and also a Uint32, -3 is only a Int32
+    - etc.
+
+Conventions on arg checking: two patterns are common:
+1. C++ functions directly exposed to the user
+2. C++ functions wrapped in a js function, only js is exposed to the user
+
+The first option requires careful checking of argument types.
+
+The second option is becoming more common. In this case the js function can
+check all the argument types are as expected (throwing an error if they are
+not), and destructure options properties to pass them as positional args (so the
+C++ doesn't have to do Object property access and presence/type checks).  C++
+can use its args fairly directly, aborting if the js layer failed to pass the
+expected types:
+
+    CHECK(args[0]->IsInt32());
+    int32_t arg0 = args[0].As<Int32>()->Value();
+
+
+- A context is an execution environment that allows separate, unrelated,
+  JavaScript code to run in a single instance of V8. The motivation for using
+  contexts in V8 was so that each window and iframe in a browser can have its
+  own fresh JavaScript environment.
+  XXX Node uses one context, mostly, does vm. create new ones? anything else?
+
+  Can get from `isolate->GetCurrentContext()`
+
+
+XXX Function vs FunctionTemplate ... wth?
+
+
+- node::Environment contains an Isolate and a Context, various other
+  information global to Node, and many convenient methods.
+  
+  It is possible to get an Environment from many v8 objects by calling
+  Environment::GetCurrent() on a v8::Isolate*, v8::Local<v8::Context>,
+  v8::FunctionCallbackInfo<v8::Value>, etc...
+
+  An Environment can be used to get an Isolate, Context, uv_loop_t, etc.
+ 
+  Commonly used convenience methods:
+  - ThrowError/TypeError/RangeError/ErrnoException/...
+  XXX why are some called Error and others called Exception?
+  - SetMethod/SetMethodNoSideEffect/SetProtoMethod/SetTemplateMethod
+  XXX difference?
+  - SetImmediate
+  - EnvironmentOptions* options(): "some" options...
+  XXX how to reach PerIsolateOptions, PerProcessOptions
+  - etc.
+
+  Contains many global strings and symbols.
+  XXX ...
+
+
+## `NODE_MODULE_CONTEXT_AWARE_INTERNAL`
+
+See: https://nodejs.org/api/addons.html#addons_context_aware_addons
+
+Called with:
+- `Local<Object> exports`: where to put exported properties, conventionally
+  called `target` in node
+- `Local<Value> module`: conventionally unused in node
+XXX what is this for? addon docs don't mention or use it
```

Comment:
It should be the same value that `module` has in CJS scripts – I think this is mostly there so that it matches the addon API more closely? We can remove it if we want, I’d say.